lib/command/src/Obelisk/Command/Preprocessor.hs

```diff
@@ -17,40 +17,39 @@ import System.FilePath (hasTrailingPathSeparator, joinPath, splitPath)
+-- Thus we must select among the packagePaths for the file we are going to parse.
 applyPackages :: FilePath -> FilePath -> FilePath -> [FilePath] -> IO ()
-applyPackages origPath inPath outPath packagePaths' = withFile outPath WriteMode $ \outFile -> do
-  -- This code is intended to be executed via ghci's -pgmF preprocessor option
-  -- The command line arguments are passed in via ghc, which dictates the first three options and meanings
-  -- In order for this code to execute, origPath must contain either a '.' character or a '/' character.
-  -- (This is to avoid the possibility of the command line syntax conflicting with another ob command)
-  -- We do have control over the remaining arguments, but they must be the same for all files.
-  -- Thus, the fourth command line argument must be "apply-packages",  which has already been handled.
-  -- We assume all the remaining arguments passed in are paths to cabal or hpack package specifications.
-  -- Thus we must select among the packagePaths for the file we are going to parse.
-
+applyPackages origPath inPath outPath packagePaths' = do
   origPathCanonical <- canonicalizePath origPath
   packagePaths <- traverse canonicalizePath packagePaths'
 
-  let takeDirs = takeWhile hasTrailingPathSeparator
-      packageDirs = sortOn (negate . length . takeDirs) $ map splitPath packagePaths
-      origDir = splitPath origPathCanonical
-      matches = [ joinPath d | d <- packageDirs, takeDirs d `isPrefixOf` origDir ]
+  let
+    takeDirs = takeWhile hasTrailingPathSeparator
+    packageDirs = sortOn (negate . length . takeDirs) $ map splitPath packagePaths
+    origDir = splitPath origPathCanonical
+    matches = [ joinPath d | d <- packageDirs, takeDirs d `isPrefixOf` origDir ]
 
-  -- So the first element of matches is going to be the deepest path to a package spec that contains
+  -- The first element of matches is going to be the deepest path to a package spec that contains
   -- our file as a subdirectory.
-
   case matches of
-    [] ->
-      hPutStrLn stderr $ "Error: Unable to find cabal information for " <> origPath <> "; Skipping preprocessor."
-    packagePath:_ -> do
-      parseCabalPackage' packagePath >>= \case
-        Left err ->
-          hPutStrLn stderr $ "Error: Unable to parse cabal package " <> packagePath <> "; Skipping preprocessor on " <> origPath <> ". Error: " <> show err
-        Right (_warnings, packageInfo) ->
-          hPutTextBuilder outFile (generateHeader origPath packageInfo)
-
-  BL.readFile inPath >>= BL.hPut outFile
+    [] -> hPutStrLn stderr $ "Error: Unable to find cabal information for " <> origPath <> "; Skipping preprocessor."
+    packagePath:_ -> parseCabalPackage' packagePath >>= \case
+      Left err ->
+        hPutStrLn stderr $ "Error: Unable to parse cabal package " <> packagePath <> "; Skipping preprocessor on " <> origPath <> ". Error: " <> show err
```

Comment:
Note that there used to be a `BL.readFile origPath >>= BL.hPut hOut` here that doesn't exist after factoring out `writeOutput`:  this was intentional in the original implementation,  otherwise I don't think GHC will see anything.  

I'm a little unclear on the exact semantics of the `-pgm*` options,  but if this change is also intentional,  then I'm guessing that

1.  the preprocessor should probably return something other than `0` as the process error code   (maybe it should do this in any case?  I dunno, it might be worth some additional experimentation/reading the code regarding how GHC actually responds to this)
2.  The messages above should be probably remove the "skipping preprocessor" language.  (Though, technically previously they were more akin to warnings than errors.)

Reply:
No this was unintentional. Good catch.